### PR TITLE
Switch to importlib-metadata

### PR DESCRIPTION
**Pull request by asottile**

### CODE REVIEW

**Comment by asottile** [comment]
This is WIP:

- [ ] https://github.com/pytest-dev/pluggy/pull/199 needs to happen first
- [ ] a few tests are failing
- [ ] figure out what to do about `pkg_resources` + `sys.path` manip and namespace packages (why do we support this? should we? does it matter in py3only world?)

Initial profiling shows this making a 20-30% improvement in startup time, I expected it to be a little better than that (more like 40-50%) but I haven't done any additional profiling yet on where I'm bleeding a bit, I suspect it *might* be due to `packaging.version` import but haven't tracked it down yet

**Comment by nicoddemus** [comment]
Thanks @asottile!

> figure out what to do about pkg_resources + sys.path manip and namespace packages (why do we support this? should we? does it matter in py3only world?)

According to [packaging namespace packages](https://packaging.python.org/guides/packaging-namespace-packages/), `pkg_resource`-style packages are still supported and not explicitly deprecated. I see 3 possible approaches:

1. Continue to support them; it doesn't seem to be a problem because we only need to call `fixup_namespace_packages` in the two `monkeypatch` and `pytester` methods which manipulate `sys.path`, so they wouldn't affect the startup time in general. This is the safest bet.

1. Drop support when we drop Python 2 in a few months. This of course needs a major release, but I'm also afraid that we can't reliably detect if `pkg-resource`-style namespace packages are being in actual use so we can at least start to emit a warning in the next releases. This is the harshest approach as may break existing and valid setups.

1. Introduce an option that enables support for `pkg-resource`-style namespace packages, defaulting to `True` in the next releases but changed to `False` in the next major release. This allows people that need the support to at least reenable the support and manifest interest in keeping the support around. This is a middle ground in order to eventually dropping the support in the future with minor disruption.

Given that we are talking about just an import and a function call in little-used functions, I'm leaning towards option 1. What do you think?

**Comment by asottile** [comment]
Seems fine whatever way.  I was **very surprised** to see the call at all as it's such a severe edgecase (you'd have to be using `pytester` to test a pytest plugin provided by a namespace package -- do any of these even exist?)

I think it also doesn't affect the subprocess-based approach :thinking:  I'll have to look at it more

**Comment by asottile** [comment]
This is now passing all of the tests, however it requires the PR of pluggy linked above.

Here's some difference in performance while using this branch:

### setup

I have 150 things installed inside pytest, including these plugins

```console
$ pip freeze | wc -l
150
$ pytest --version
This is pytest version 4.4.1, imported from /tmp/paasta/venv/lib/python3.6/site-packages/pytest.py
setuptools registered plugins:
  pytest-asyncio-0.10.0 at /tmp/paasta/venv/lib/python3.6/site-packages/pytest_asyncio/plugin.py
  hypothesis-3.71.10 at /tmp/paasta/venv/lib/python3.6/site-packages/hypothesis/extra/pytestplugin.py
```

### before

(best of 5)

```console
$ time pytest --help  >& /dev/null

real	0m0.555s
user	0m0.499s
sys	0m0.056s
```

### after (~21% improvement)

(best of 5)

```console
$ time pytest --help  >& /dev/null

real	0m0.437s
user	0m0.413s
sys	0m0.024s
```

There's probably even more to save here but I'd need to do additional profiling on that

**Comment by asottile** [comment]
(this will have to re-target `features` instead -- wanted to get feedback on this first though!)

**Comment by nicoddemus** [comment]
Added this one to the `4.5` milestone then. 👍

**Comment by asottile** [comment]
> 
> 
> Added this one to the `4.5` milestone then. 👍

sweet, I'll rebase / retarget this when pluggy is released 🎉

**Comment by asottile** [comment]
rebased, here's the results using pytest-dev/pluggy#215

```
====== 2265 passed, 131 skipped, 11 xfailed, 1 warnings in 283.17 seconds ======
```

**Comment by asottile** [comment]
@nicoddemus @blueyed this should be ready for review 🎉

File: changelog/5063.feature.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Switch from ``pkg_resources`` to ``importlib-metadata`` for entrypoint detection for improved performance and import time.
```
Comment (nicoddemus) [inline changelog/5063.feature.rst]:
```suggestion
Switch from ``pkg_resources`` to ``importlib-metadata`` for entrypoint detection for improved performance and import time.

As a result of this, ``importlib-metadata`` is a new dependency.
```

Comment (nicoddemus) [reply changelog/5063.feature.rst]: Hmmm it can be argued that this is obvious from the initial description, so feel free to discard this. Up to you. 👍

Comment (asottile) [reply changelog/5063.feature.rst]: going to just leave it then -- thanks for the review!

**Comment by asottile** [comment]
when do we want to make the 4.6 release? I can run it if we'd like :D

**Comment by nicoddemus** [comment]
I think probably after @blueyed can get to https://github.com/pytest-dev/pytest/pull/5307#discussion_r287599922

**Comment by nicoddemus** [comment]
> I can run it if we'd like 

Sure, please go ahead btw! 😁

File: testing/test_assertion.py [hunk 2]
```diff
@@ -137,12 +137,12 @@ def test_foo(pytestconfig):
...
         """Test that we match files correctly when they are marked for rewriting (#2939)."""
         contents = {
-            "conftest.py": """
+            "conftest.py": """\
                 pytest_plugins = "ham"
             """,
             "ham.py": "",
             "hamster.py": "",
-            "test_foo.py": """
+            "test_foo.py": """\
```
Comment (blueyed) [inline testing/test_assertion.py]: Was the escaped newline required for some reason?

Comment (asottile) [reply testing/test_assertion.py]: no but it makes the line numbers add up and makes my editor less angry about tqs

Comment (blueyed) [reply testing/test_assertion.py]: tqs?

**Comment by blueyed** [comment]
As for 4.6 I've created a milestone: https://github.com/pytest-dev/pytest/milestone/28
Will move the Pdb wrapper class tomorrow.

File: testing/test_assertion.py (hunk 2 above)
Comment (asottile) [reply testing/test_assertion.py]: triple quoted strings

**Comment by asottile** [comment]
@blueyed are these blockers for the release? I'd really like to get this out so we can start cleaning things up.  If they're non-essential / bugfixes they can always come later right?

**Comment by blueyed** [comment]
https://github.com/pytest-dev/pytest/pull/5322 should be in to avoid conflicts later.
https://github.com/pytest-dev/pytest/issues/5301 is a regression in 4.5 already it seems (IIRC).
And https://github.com/pytest-dev/pytest/pull/5313 changes CI in a way that should also not diverge (but could maybe go into master then, i.e. 4.6.1)

**Comment by asottile** [comment]
I'd like to at least drop #5301 / #5313 from blocking 3.6 -- they're bugfixes that can happen at a later date -- I don't know enough about #5322 to make a call there

**Comment by blueyed** [comment]
Yeah, go ahead and remove them then - also #5322 then maybe.
But I'm not sure why there's a need to rush this?

**Comment by asottile** [comment]
oh, the plan was to drop 2.7 / 3.4 by end of May which is fast approacing :)

**Comment by nicoddemus** [comment]
> Yeah, go ahead and remove them then - also #5322 then maybe.

When ahead and removed #5313, it does seem like it can be done afterwards. 

I agree that #5322 might cause unwanted conflicts later, it would be nice to get that done before 4.6. Any blockers there?

**Comment by blueyed** [comment]
> #5322 
> Any blockers there?

1. missing coverage (want to test that it re-uses the class), but that's easy
2. cleaning up further, like you suggested
Let's rather discuss this there further - I plan to get back to it tomorrow or on the weekend otherwise.

**Comment by blueyed** [comment]
But apart from that we should / need do something about the failures on features anyway, before merging it into master.
Feel free to pick up #5334 in this regard.

**Comment by nicoddemus** [comment]
> Feel free to pick up #5334 in this regard.

I will make some time for it tomorrow, thanks for narrowing it down. 👍

**Comment by lvt-mm** [comment]
I'm surprised that you'd pull in a dependency marked as alpha quality in a package marked as stable. I haven't looked into any of the history or motivation of this change, but in general I'd like my stable dependencies to have stable sub-dependencies.

**Comment by nicoddemus** [comment]
@lvt-mm that's a valid point, but I was under the impression that `importlib-metadata` was stable, but you are right that it is marked as "Alpha" on PyPI.

@asottile @jaraco @brettcannon would you like to comment?

**Comment by asottile** [comment]
It's going to be in cpython 3.8, the api has been stable :man_shrugging:

**Comment by nicoddemus** [comment]
It was my impression as well. Perhaps `importlib-metadata` should be changed to `Beta` or `Production` then? 🤔

**Comment by brettcannon** [comment]
@warsaw and @jaraco will know best, but I believe it's just an oversight in not updating the metadata.

**Comment by warsaw** [comment]
The API is now stable.  We'll release `importlib_metadata` 1.0 and update its, um metadata when Python 3.8rc1 is released.

**Pull request merged**